homeassistant/components/hassio/update.py

```diff
@@ -164,9 +166,12 @@ async def async_install(
         **kwargs: Any,
     ) -> None:
         """Install an update."""
+        supervisor_client = get_supervisor_client(self.hass)
```

Comment:
I'd store the client on the entity instance in the init method.

Reply:
So I couldn't use the `__init__` method to do that for `AddonManager` as it seemed to create a race condition with the `HassIO` object being created and added to the hass object. That's why I went with the `property` that caches it on first access. I wasn't sure if that was only an issue in unit tests or something that would occur live as well though.

Is it safe to do that here? Or better to just go with a `property` like this:
https://github.com/home-assistant/core/blob/494511e099772722fc9daee15682a8abdf9bc939/homeassistant/components/hassio/addon_manager.py#L116-L121

Reply:
Yea as with before doing this in `__init__` causes all the unit tests to fail:
```py
        self._supervisor_client: SupervisorClient | None = get_supervisor_client(self.hass)
```
But caching on first access to a private property works fine. I'm guessing this is not a real issue but rather one with how our mocking is setup. Because the fixture that mocks that `get_supervisor_client` isn't in place when all the objects are being initialized by a different fixture.

Assuming this isn't an issue on a live system I cant try and do some refactoring here to make this work if you think it is preferrable over the private property above. Since we were ok with that approach in `AddonManager` I assume its ok here as well.

Reply:
I don't think we should store the client on the HassIO instance at all. I think we should use the singleton decorator on the function to get the client.

Reply:
<https://github.com/home-assistant/core/blob/2f60395ba9eaf9fb0a0b390267275ba0994a9c35/homeassistant/helpers/singleton.py#L27>

Reply:
Ok that worked. It took a bit though. I don't really like how I had to patch mock every place I imported `get_supervisor_client` in the `supervisor_client` fixture to make it functional but nothing else seemed to work. If there's a better way to do that let me know, patching mocking only `homeassistant.components.hassio.handler.get_supervisor_client` did not seem to be enough. But at least the runtime code is nicer which is more important.

Reply:
We need to patch where the name is used, so I think this is expected and ok.